src/System.Text.Http/System/Text/Http/Parser/HttpParser.cs

```diff
@@ -107,271 +105,151 @@ public HttpParser(bool showErrorDetails)
+        private void ParseRequestLine<T>(ref T handler, ReadOnlySpan<byte> data) where T : IHttpRequestLineHandler
         {
-            // Get Method and set the offset
-            var method = HttpUtilities.GetKnownMethod(data, length, out var offset);
+            ReadOnlySpan<byte> customMethod = default;
 
-            ReadOnlySpan<byte> customMethod =
-                method == Http.Method.Custom ?
-                    GetUnknownMethod(data, length, out offset) :
-                    default;
+            // Get Method and set the offset
+            if (!HttpUtilities.GetKnownMethod(data, out Http.Method method, out int offset))
+            {
+                customMethod = GetUnknownMethod(data, out offset);
+            }
 
-            // Skip space
+            // Skip the space after the method
             offset++;
 
-            byte ch = 0;
-            // Target = Path and Query
-            var pathEncoded = false;
-            var pathStart = -1;
-            for (; offset < length; offset++)
+            if (offset >= data.Length)
             {
-                ch = data[offset];
-                if (ch == ByteSpace)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Empty path is illegal
-                        RejectRequestLine(data, length);
-                    }
+                // Need some path data
+                RejectRequestLine(data);
+            }
 
-                    break;
-                }
-                else if (ch == ByteQuestionMark)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Empty path is illegal
-                        RejectRequestLine(data, length);
-                    }
+            bool pathEncoded = false;
+            ReadOnlySpan<byte> target = data.Slice(offset);
+            int pathEnd = target.IndexOfAny(ByteSpace, ByteQuestionMark);
+            if (pathEnd < 1 || pathEnd > target.Length - 1)
+            {
+                // Cant start or end with space/? or eat the entire target
+                RejectRequestLine(data);
+            }
 
-                    break;
-                }
-                else if (ch == BytePercentage)
-                {
-                    if (pathStart == -1)
-                    {
-                        // Path starting with % is illegal
-                        RejectRequestLine(data, length);
-                    }
+            ReadOnlySpan<byte> path = target.Slice(0, pathEnd);
 
-                    pathEncoded = true;
-                }
-                else if (pathStart == -1)
-                {
-                    pathStart = offset;
-                }
+            int escapeIndex = path.IndexOf(BytePercentage);
+            if (escapeIndex == 0)
+            {
+                // Can't start with %
+                RejectRequestLine(data);
```

Comment:
consider `goto Error` for all the error cases.

Reply:
Tried, it made it worse.